src/testing/byzantine.rs

```diff
@@ -0,0 +1,409 @@
+use codec::{Decode, Encode};
+
+use log::{debug, error};
+use tokio::time::Duration;
+
+use futures::{
+    channel::{mpsc::UnboundedReceiver, oneshot},
+    FutureExt, StreamExt,
+};
+
+use std::collections::HashMap;
+
+use crate::{
+    member,
+    member::Config,
+    nodes::NodeMap,
+    signed::Signed,
+    testing::mock::{
+        node_ix_to_peer, AlertHook, Data, DataIO, Hasher64, KeyBox, Network, NetworkHub, PeerId,
+        Signature, Spawner,
+    },
+    units::{ControlHash, FullUnit, PreUnit, SignedUnit, UnitCoord},
+    Hasher, Network as NetworkT, NetworkCommand, NetworkEvent, NodeCount, NodeIndex, OrderedBatch,
+    SessionId, SpawnHandle,
+};
+
+type Hash64 = <Hasher64 as Hasher>::Hash;
+
+use crate::member::{ConsensusMessage, ConsensusMessage::*};
+
+struct MaliciousMember<'a> {
+    node_ix: NodeIndex,
+    n_members: NodeCount,
+    threshold: NodeCount,
+    session_id: SessionId,
+    forking_round: usize,
+    round_in_progress: usize,
+    keybox: &'a KeyBox,
+    network: Network,
+    unit_store: HashMap<UnitCoord, SignedUnit<'a, Hasher64, Data, KeyBox>>,
+}
+
+impl<'a> MaliciousMember<'a> {
+    fn new(
+        keybox: &'a KeyBox,
+        network: Network,
+        node_ix: NodeIndex,
+        n_members: NodeCount,
+        session_id: SessionId,
+        forking_round: usize,
+    ) -> Self {
+        let threshold = (n_members * 2) / 3 + NodeCount(1);
+        MaliciousMember {
+            node_ix,
+            n_members,
+            threshold,
+            session_id,
+            forking_round,
+            round_in_progress: 0,
+            keybox,
+            network,
+            unit_store: HashMap::new(),
+        }
+    }
+
+    fn send_network_command(&mut self, command: NetworkCommand) {
+        if let Err(e) = self.network.send(command) {
+            panic!("Failed to send network command {:?}.", e);
+        }
+    }
+
+    fn pick_parents(&self, round: usize) -> Option<NodeMap<Option<Hash64>>> {
+        // Outputs a parent map if there are enough of them to create a new unit.
+        let mut parents = NodeMap::new_with_len(self.n_members);
+        if round == 0 {
+            return Some(parents);
+        }
+        let mut count = 0;
+        let mut our_unit_available = false;
```

Comment:
Maybe check it straight away? I'm not sure how often pick_parents misses, though.

Reply:
Added this check -- it avoids the loop now, so it's better for sure.